### ISSUE TITLE
Code analysis with Valgrind and Infer

### ISSUE THREAD

**Issue by mindhells** [fract4d/gnofract4d#109]
I'm working on this [branch](https://github.com/HyveInnovate/gnofract4d/commits/examples-code-analysis). There I've added [Valgrind](https://valgrind.org/) and [Infer](https://fbinfer.com/) tools for the C++ [examples](https://github.com/fract4d/gnofract4d/tree/master/examples). 

Examples readme has also been updated there explaining how to run the examples using Valgrind.
No memory leaks detected so far in the C++ model... and that's the reason me opening this issue: is that true or am I doing something wrong?

**Comment by edyoung** [comment]
Cool! You could try introducing av Leah d deliberately and see if it is reported...

**Comment by mindhells** [comment]
Good point. I guess I assumed there should be some, maybe because I underestimated the current implementation.

I know this analysis won't prove much about Gnofract4d running in GUI mode. For in that case, there's "objects" being reused or disposed depending on the use case (changing the formula o zooming in/out), and to find if there's leaking there I suppose you'll need python specific tools like [Pympler](https://github.com/pympler/pympler).

I still find tricky, when creating Python extensions, to deal with [reference counting](https://docs.python.org/3/c-api/refcounting.html). I wonder if there's a manageable way to find leaking problems there.

**Comment by edyoung** [comment]
Re: leaking problems, I had a crappy manual way to deal with that. If you enable `#DEBUG_ALLOCATION=1` you get some debug spew which logs when each python-referenced object is created or destroyed. You run the program, record the trace, then sort the output. If there is a line with `0x12345 : FOO : CTOR` not followed by a `0x12345 : FOO : DTOR` then you leaked a FOO. Crude, but found a number of bugs that way.

**Comment by mindhells** [comment]
No time to try it out yet, but I'll drop here this interesting piece of info, just in case someone runs into this issue: https://pythonextensionpatterns.readthedocs.io/en/latest/debugging/valgrind.html#building-python-for-valgrind-label
